src/mcts/node.cc

```diff
@@ -318,12 +301,67 @@ void NodeTree::MakeMove(Move move) {
   history_.Append(move);
 }
 
+// This function brute force examines the entire tree to recalculate the already
+// searched depth. At first glance this might seem excessive, but, even aside
+// from the memory savings of not caching depths in Nodes, this is actually in
+// sum a computational savings as well. The simple reason is that for the cached
+// depths, they are recalculated n times for *every* node, where n is the depth.
+// This recalculates *once* per node (and only the reused nodes at that), so
+// this replaces O(N log(N)) with 2*O(N) computation (N is the number of nodes,
+// n ~ log(N)). And, as if that wasn't cool enough, this actually moves some of
+// the computation out of the otherwise-critical search code/cpu time, and moves
+// it into the pre-search initialization which doesn't count against move-search
+// time. The only drawback is that this computational effort is changed from
+// spread-over-the-search to all-at-once-in-initializtion, but it's a very
+// simple and efficient loop, and even with dozens of millions of nodes, should
+// only take a second or so (and UCI sends an `isready` command between
+// `position` and `go` anyways).
+void NodeTree::RecalculateDepth() {
+  // This is the iterative translation of the most obvious and simple recursive
+  // implementation of a brute force traversal of the entire tree. Basically the
+  // two choices for this are depth first and breadth first; the former is far
+  // more obvious from a recursive standpoint, and also has the advantage that
+  // the intermediate storage need only be logarithmically long relative to the
+  // number of nodes. (The latter, using a queue rather than a stack, would have
+  // an intermediate length much closer to linear in the number of nodes rather
+  // than logarithmic.)
+  //
+  // The recursive implementation looks like this:
+  // RecalculateDepth(node, depth) {
+  //   process node at depth
+  //   for each child of node:
+  //     RecalculateDepth(child, depth+1)
+  // }
+  std::stack<Node*> stack;
+  cumulative_depth_ = max_depth_ = 0;
+
+  stack.push(current_head_);
+
+  while (!stack.empty()) {
```

Comment:
Traversing the whole tree just to recalculate depth? No, that's not a way to go.
Just reset all records to zero at the beginning of the search and that's it.

Reply:
It's less work than is done in current code, and it's not in any time managed code, and takes milliseconds. Why not? See my stream, where I'm running this PR now.

I strongly believe that making the UCI printed depth ignore the `NodeTree` will lead to inconsistent and confusing output. It would mean that the depth printed for a given position analysis changes wildly depending on the order of commands entered, even if the exact nodes searched are literally identical.

Reply:
Here's another way to do it: if it's not okay for my code to do O(N) work in this PR, then why is it okay for your code *currently in master* to do O(N log N) work? This PR quite literally reduces the total operations relating to depth computation.

Reply:
I guess I'd be okay with adding an option to disable depth computation of reused trees, but I'd recommend for TCEC games that reused trees must be included, otherwise the output will be grossly inconsistent and confusing. All those 2-second-insta-moves due to smart pruning on reused trees will indicate a depth of ~0, which is obviously incorrect when 2M nodes are reported.

Reply:
As an example, from my test which is ongoing:

```
1. b3 book b6 book 
2. Bb2 book Bb7 book 
3. c4 +0.12/16 453s c5 -0.05/15 622s 
4. Nf3 +0.11/17 716s Nf6 -0.07/16 1179s 
5. e3 +0.14/18 749s e6 -0.05/18 1403s 
6. Be2 +0.15/17 163s Be7 -0.05/18 338s 
7. O⁠-⁠O +0.18/17 1076s O⁠-⁠O -0.07/18 1634s 
8. Nc3 +0.17/15 1284s d5 -0.08/19 1028s 
9. cxd5 +0.10/17 2726s Nxd5 -0.08/17 161s 
10. Nxd5 +0.10/16 75s Bxd5 -0.08/16 73s 
11. d4 +0.28/19 3098s Nd7 -0.08/20 3118s 
12. Rc1 +0.28/20 2567s h6 -0.08/19 186s 
13. Bc4 +0.40/19 2754s Be4 -0.14/18 2639s 
14. Qe2 +0.34/19 2251s Qc7 -0.14/17 232s 
15. Rfd1 +0.53/16 2308s Rad8 -0.13/17 315s 
16. dxc5 +0.49/18 2244s bxc5 -0.27/17 2237s 
17. h3 +0.59/16 2012s Nb6 -0.26/15 58s 
18. Be5 +0.63/21 2301s Qc8 -0.22/19 733s 
19. Ba6 +0.66/21 2281s Rxd1+ -0.25/20 774s 
20. Rxd1 +0.72/21 1339s Qc6 -0.26/19 166s 
21. Bb5 +0.75/21 617s Qb7 -0.28/19 484s 
22. Ba6 +0.77/21 514s Qc6 -0.32/19 621s 
23. Bg3 +0.84/22 2642s Qa8 -0.51/17 2725s 
24. Ne5 +0.88/22 2403s Rd8 -0.54/17 1228s 
25. Rxd8+ +0.86/23 745s Qxd8 -0.69/21 2076s 
26. Qg4 +0.88/24 776s Bd5 -0.66/21 2455s 
27. e4 +1.08/27 2463s Ba8 -0.70/25 2533s 
28. Qh5 +1.08/25 26s Qe8 -0.69/23 63s 
29. Bb5 +1.03/25 1345s Qf8 -0.70/22 223s 
30. Qf3 +1.04/23 54s Bd6 -0.73/22 405s 
31. Qd3 +1.03/23 316s Bc7 -0.73/20 73s 
32. a4 +1.04/21 48s Qe7 -0.66/19 3348s 
33. Ba6 +1.19/23 1002s Qg5 -0.66/27 2379s 
34. Nxf7 +1.33/33 2275s ```

Notice how the depths are slightly different for each move, reflecting the shape of the tree (all moves are at fixed 2M nodes, smart pruning disabled)? Now notice the tree reuse that causes sometimes super fast moves (by comparison) at 10 times or greater faster than the longest move. Those tree reuse moves would have depths of 5 or 10 instead of 18-25, and we would no longer be able to compare the depths of each and every move.

Reply:
Here's a recording of several large tree reuses in a row.

https://clips.twitch.tv/VenomousFunWormDoggo

Reply:
Cscuile on Discord today confirmed that there is not any noticeable added time in starting searches, even with trees with several million nodes.

> Cscuile: @Dubslow I'm not noticing any significant differences.

> Dubslow: but it doesn't lag before starting its search? that's the part im worried about
Cscuile: Nope it doesnt for me